src/main/java/com/google/cloud/genomics/dataflow/readers/ReadStreamer.java

```diff
@@ -0,0 +1,144 @@
+ * the License.
+ */
+package com.google.cloud.genomics.dataflow.readers;
+
+import com.google.api.services.genomics.model.ReadGroupSet;
+import com.google.api.services.genomics.model.SearchReadGroupSetsRequest;
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.genomics.grpc.Channels;
+import com.google.cloud.genomics.utils.Contig;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+import com.google.cloud.genomics.utils.Paginator;
+import com.google.common.base.Function;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.Lists;
+import com.google.genomics.v1.Read;
+import com.google.genomics.v1.StreamReadsRequest;
+import com.google.genomics.v1.StreamReadsResponse;
+import com.google.genomics.v1.StreamingReadServiceGrpc;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Class with tools for streaming reads using gRPC within dataflow pipelines.
+ */
+public class ReadStreamer {
+
+  /**
+   * Gets ReadGroupSetIds from a given datasetId using the Genomics API.
+   */
+  public static List<String> getReadGroupSetIds(String datasetId, GenomicsFactory.OfflineAuth auth)
+      throws IOException, GeneralSecurityException {
+    List<String> output = Lists.newArrayList();
+    Iterable<ReadGroupSet> rgs = Paginator.ReadGroupSets.create(
+        auth.getGenomics(auth.getDefaultFactory()))
+        .search(new SearchReadGroupSetsRequest().setDatasetIds(Lists.newArrayList(datasetId)),
+            "readGroupSets/id,nextPageToken");
+    for (ReadGroupSet r : rgs) {
+      output.add(r.getId());
+    }
+    if (output.isEmpty()) {
+      throw new IOException("Dataset " + datasetId + " does not contain any ReadGroupSets");
+    }
+    return output;
+  }
+
+  /**
+   * Constructs a StreamReadsRequest for a readGroupSetId, assuming that the user wants all
+   * to include all references.
+   */
+  public static StreamReadsRequest getReadRequests(String readGroupSetId) {
+    return StreamReadsRequest.newBuilder()
+        .setReadGroupSetId(readGroupSetId)
+        .build();
+  }
+
+  /**
+   * Constructs a StreamReadsRequest for a readGroupSetId and a set of user provided references.
+   */
+  public static List<StreamReadsRequest> getReadRequests(final String readGroupSetId,
+      String references) {
+    final Iterable<Contig> contigs = Contig.parseContigsFromCommandLine(references);
+    return FluentIterable.from(contigs)
+        .transformAndConcat(new Function<Contig, Iterable<Contig>>() {
+            @Override
+            public Iterable<Contig> apply(Contig contig) {
+              return contig.getShards(100000000L);
```

Comment:
Yeah, I don't think it should be hardcoded, but I'm not sure of a good
general purpose heuristic for determining an appropriate size of shard.
Keep in mind these are streaming read requests, so we don't actually want
too many shards operating at the same time. It's a balance between "small
enough" that Dataflow can re-try automatically for you without duplicating
a lot of data, and "large enough" that you get streaming throughput
benefits outweighing the startup costs.  In absence of a better algorithm,
I thought we'd just start with this. Maybe see what comes up in your
modification of the other samples for an appropriate function?

On Mon, Jun 22, 2015 at 1:02 PM, Nicole Deflaux <notifications@github.com>
wrote:

> In
> src/main/java/com/google/cloud/genomics/dataflow/readers/ReadStreamer.java
> <https://github.com/googlegenomics/dataflow-java/pull/99#discussion_r32976704>
> :
>
> > +    return StreamReadsRequest.newBuilder()
> > +        .setReadGroupSetId(readGroupSetId)
> > +        .build();
> > +  }
> > +
> > +  /**
> > +   * Constructs a StreamReadsRequest for a readGroupSetId and a set of user provided references.
> > +   */
> > +  public static List<StreamReadsRequest> getReadRequests(final String readGroupSetId,
> > +      String references) {
> > +    final Iterable<Contig> contigs = Contig.parseContigsFromCommandLine(references);
> > +    return FluentIterable.from(contigs)
> > +        .transformAndConcat(new Function<Contig, Iterable<Contig>>() {
> > +            @Override
> > +            public Iterable<Contig> apply(Contig contig) {
> > +              return contig.getShards(100000000L);
>
> Is it correct that the should always be hardcoded?
>
> If so, this hardcoded value here and in VariantStreamer could be
> refactored as a constant in com.google.cloud.genomics.utils.Contig
>
> But I think it should not be hardcoded - let me know if I am missing
> something. For example when used in the context of a Dataflow job running
> on human chr1 with 248,956,422 base pairs, we will only have three shards,
> so only a max of three Dataflow workers will be able to do any work during
> the retrieve-data-via-grpc stage of the pipeline.
>
> See also:
>
>    -
>    https://github.com/googlegenomics/dataflow-java/blob/master/src/main/java/com/google/cloud/genomics/dataflow/utils/GenomicsDatasetOptions.java#L128
>    -
>    https://github.com/googlegenomics/utils-java/blob/master/src/main/java/com/google/cloud/genomics/utils/Contig.java#L43
>    -
>    https://github.com/googlegenomics/dataflow-java/blob/master/src/main/java/com/google/cloud/genomics/dataflow/utils/GenomicsDatasetOptions.java#L55
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/googlegenomics/dataflow-java/pull/99/files#r32976704>.
>